### ISSUE TITLE
[UX] move the clear log messages button to bottom of dblog messages page

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#2353]
The question here is: **Where should we put this "Clear Log Messages" Button?** 

There are several UX improvements we are addressing here:
* avoid accidental clicking/deleting FOR ADMINS + DEVS
* prevent confusion from too many available options above the listing FOR ADMINS + DEVS
* reduce the amount of scrolling needed to read the listing FOR ADMINS + DEVS
* make this interface more consistent with the rest of core FOR ADMINS + DEVS
* reduce #of clicks to delete messages ? FOR DEVS (only)

We will eventually want to remove the fieldset around the filters to reduce the number of clicks needed to filter the listing, too. (This will also make this interface more consistent with the rest of core). Our decision here should take that into account. 

I indicated the ROLE of the person completing each task above because in Backdrop we put the needs of ADMINS above the needs of the needs of DEVeloperS. 

Options as follows.

1) Leave it where it is, at the top of the page, buried in a fieldset:
<img width="1232" alt="Screen Shot 2020-02-20 at 10 57 54 AM" src="https://user-images.githubusercontent.com/397895/74968789-45593a80-53d0-11ea-8ea9-e9d884c7cc92.png">

2) Move it to the bottom of the page, but NOT in a fieldset: 
<img width="1606" alt="screen shot 2018-04-22 at 11 22 12 am" src="https://user-images.githubusercontent.com/6763671/39096608-56bc28d8-4620-11e8-81b2-586e47ac5e8e.png">

3) Move it to another page, accessed by a tab, like Drupal 8 did:
![screenshot-drupal8](https://user-images.githubusercontent.com/842066/73439801-7fdf3280-4350-11ea-89cd-980092d7bf8c.png)

4) Combine the two fieldsets for filtering the current list, and deleting all items to save space
(no screenshot available)

5) move that button to the Log settings page: `/admin/config/development/logging`?

### Objections to 1:

> The clear log messages is inside a fieldset. I understand that we don't want people clicking this by accident, but I'm not sure a fieldset is the way to deal with the problem

> I don't know of anywhere else in core where we make users go to the listing page in order to delete items. The delete option is always on another page/tab.

### Objections to option 2

> [DEVS will] need to scroll down this page 

> it might seem like a "save" button or something people are supposed to do here, and might increase the probability of people doing it by accident.

### Objections to option 3

> After doing UX reviews/studies, D8 have decided to remove all tabs and convert them to action links

> introduction of [a new] UI pattern of having an action live as a primary/secondary tab. It just does not feel right UX-wise to me. 

> There have been extensive discussions and user studies in Drupal-land, concluding that things that are actions (such as the "delete" operation), should be buttons/links and not tabs; tabs are being reserved for navigational operations.

### Objections to option 4

> taking [the Clear Log Messages button] out of the fieldset and then having it anywhere near the buttons for the filters (above, below, to the right) would also be confusing [for ADMINS and DEVS]

> If the button can be moved away from the top of the page (above what you're trying to access 99%* of the time), then I agree that a fieldset is not needed. But as long as the button's above the content, it should be in a collapsed fieldset.

### Objections to option 5

>  it's convenient for DEVS to have here - right where you are anyway when you are debugging

> Definitely wrong place with too long path, we need clear log button on same screen where we read it.

---

Original Issue

---

The 'Recent log messages' page has several UI patterns that are unusual for core. 

The clear log messages is inside a fieldset. I understand that we don't want people clicking this by accident, but I'm not sure a fieldset is the way to deal with the problem, especially since it separates the filters at the top of the page with the list the filters are acting on, below the fieldset.  How about we 1) remove the fieldset and 2) move the button to the bottom of the page. 

~~First, the filters for the log messages are hidden away in a fieldset. This means you have to open the fieldset to filter the messages.  Second, the `Severity` filter is a multi-select widget, which is nearly unusable. This should be replaced with checkboxes instead.~~ See https://github.com/backdrop/backdrop-issues/issues/3051 for these issues

----
~PR: https://github.com/backdrop/backdrop/pull/2142~
PR by @opi: https://github.com/backdrop/backdrop/pull/2148 
Alternate PR by @indigoxela: https://github.com/backdrop/backdrop/pull/3048 (closed in favor of a contrib module)
Alternate PR by @klonos: https://github.com/backdrop/backdrop/pull/3055

**Comment by serundeputy** [comment]
This PR: https://github.com/backdrop/backdrop/pull/2142

Addresses the first and third bullets:
  * removes fieldsets
  * moves the clear messages to the bottom and adds `button-danger` class

The second bullet use checkboxes is not addressed:
  * I tried to use checkboxes, but I could not get the connection between type and severity to persist with checkboxes.
  * I think this improves the UX quite a bit, and we should hold off on the second bullet and address that UX issue when we get the `select2` library in core

<img width="1606" alt="screen shot 2018-04-22 at 11 22 12 am" src="https://user-images.githubusercontent.com/6763671/39096608-56bc28d8-4620-11e8-81b2-586e47ac5e8e.png">

**Comment by findlabnet** [comment]
When I open this page, I'd like to see logs itself, but not filters/buttons occupying a large part of the page (same as in D8?).
What's wrong with fieldsets? And yes, I use filters everyday, same as "clear log messages" button.
IMHO, we can combine two fieldses to one, something like "Manage log messages" and get even more space on screen for messages.

**Comment by serundeputy** [comment]
Sounds like @findlabnet and @jenlampton have conflicting ideas about what is best for this page.

Combining into a single fieldset will be tricky as they are two completely different forms being added to the page with a `$build` variable.

Talk it over and see if we can come to consensus.

**Comment by jenlampton** [comment]
Maybe we should get a ux designer to weigh in?

**Comment by BWPanda** [comment]
I'm not a UX person (just a regular developer)...

I like moving the 'clear' button to the bottom (and out of a fieldset), but I don't like the space the filters in the screenshot now take up vertically.

If select2 is close to being added to core, I'd suggest leaving the filters in a fieldset until they can utilise that instead.

**Comment by jenlampton** [comment]
> If select2 is close to being added to core, I'd suggest leaving the filters in a fieldset until they can utilise that instead.

What if it's not? Should we hold off on this change until it is?

**Comment by jenlampton** [comment]
Sounds good to me. 
Rescoping this issue to focus on moving the button for 1.10. I moved the filters fixes to https://github.com/backdrop/backdrop-issues/issues/3051 which now depends on https://github.com/backdrop/backdrop-issues/issues/1800

**Comment by opi** [comment]
PR https://github.com/backdrop/backdrop/pull/2148 is a reroll of https://github.com/backdrop/backdrop/pull/2142 with a reduced scope.

**Comment by findlabnet** [comment]
Thanks @opi, but it is still not clear to me what we have achieved with this change.

**Comment by opi** [comment]
by spliting the 2 forms (filter & reset), we reduce clutter for admin user ; Also, action buttons are often on page bottom, this is a common UX pattern.

**Comment by findlabnet** [comment]
> common UX pattern

Do you like an idea to scroll down 2-3 screen heights any time you need to clear bunch of already unuseful log messages?
Common UX pattern for Drupal admin - looking for this button on top of page, or I wrong?

**Comment by jenlampton** [comment]
With the button where it was, it also separated the filters from the results that were filtered, which was a little jarring. 

This change is in preparation for moving the filters out of their fieldset to bring them more in line with how the filters look and work on the other admin pages. 

I don't have strong feelings about having the button at the top of the page (as opposed to the bottom) -- though it was convenient (for me) to have it at the top. I do worry that taking it out of the fieldset and then having it anywhere near the buttons for the filters (above, below, to the right) would also be confusing.

**Comment by opi** [comment]
> Common UX pattern for Drupal admin [...]

Backdrop is a good place to fix some odd drupalism, and become a great product on its own <3

**Comment by laryn** [comment]
**Side note** on the UX pattern in question: I was struck the other day after editing a View and then a Layout that the action buttons are not always consistently placed. (top right / bottom left respectively, although bottom does seem more common in the rest of the interface)

**Comment by findlabnet** [comment]
My question is:
>Do you like an idea to scroll down 2-3 screen heights any time you need to clear bunch of already unuseful log messages?

This is like to UX improvement?

**Comment by jenlampton** [comment]
@laryn I think thee's an issue somewhere about Views' funky submit-button location, but if not can you create one?

@findlabnet there are several UX improvements we are addressing here 1) bringing the filters closer to the form 2) eventually, getting the filters out of the fieldset, which 3a) makes this interface more consistent with the rest of core and 3b) reduces the number of clicks needed to accomplish the goal of filtering this list. 

I think your point is that there is *also* a UX regression in moving the button to the bottom of the page, for the cases when this log is full of messages that you would like to delete. 

In my experience, most of the time I am on this page I do not need the `Clear log messages` button, so I personally won't mind scrolling to the bottom of the page when I do need it. 

This is probably the same reasoning behind burying the button it in a fieldset: In most cases, that extra click is not needed, because the button isn't needed.

**Comment by laryn** [comment]
@jenlampton I found [this](https://github.com/backdrop/backdrop-issues/issues/541#issuecomment-68689302) and [this](https://github.com/backdrop/backdrop-issues/issues/566) RE: views submit button location.

**Comment by findlabnet** [comment]
@jenlampton, thanks for clarifying, but as I mentioned above, on development stage I personally use this button very intensively and **this is only case when this button is needed**.
So, for me moving button to bottom means UX regression, but maybe it is for me only.

**Comment by jenlampton** [comment]
@larn I've created https://github.com/backdrop/backdrop-issues/issues/3054 

> on development stage I personally use this button very intensively [...] So, for me moving button to bottom means UX regression, but maybe it is for me only.

@findlabnet It's not only you, but in Backdrop we put the needs of developers below the needs of site architects, and below the needs of end-users. (see https://backdropcms.org/philosophy#importance)

**Comment by findlabnet** [comment]
Maybe I lost difference between philosophy and religion, but I don't see site architects (BTW, who is site architect in this context, maybe site buider?) or  end-users  as users of this button.
Anyway, it is just my opinion only.

**Comment by jenlampton** [comment]
> but I don't see site architects or end-users as users of this button.

Yes! That's the point exactly. They don't need this button, so it doesn't matter if we move it "out of the way". Those people need the list of log messages, and the filters for them, which is why we want those things to come at the top of the page :)

**Comment by findlabnet** [comment]
@jenlampton thank you! I got the point. ;)
So, when this changes is merged to mainstream - just need build contrib to get it back for low-priority users.

**Comment by jenlampton** [comment]
exactly!  Maybe the devel module could do it?

**Comment by findlabnet** [comment]
Moreover, I think this button should be hidden for any user except #1, or have some exclusive permissions.

**Comment by opi** [comment]
@findlabnet added some `#weight` key to the `dblog_overview()` page callback to ease the order changes in contrib or custom code. (https://github.com/backdrop/backdrop/pull/2148/files)

**Comment by jenlampton** [comment]
> I think this button should be hidden for any user except uid 1, or have some exclusive permissions.

That might be over-enginering things a little. It's not dangerous to provide this button to non-developer users... Eh, I could probably be convinced either way! :)

**Comment by opi** [comment]
why not having specific permissions ; @findlabnet could you open another issue ?

**Comment by findlabnet** [comment]
@opi - done: #3056

**Comment by docwilmot** [comment]
I dont think the button at the bottom makes much sense. Most of the time all the action youre interested in is at the top: the latest log messages. 
And this filter isnt same as other core filters because of the multiple, so I dont think the same UI pattern will be easy.
My suggestion would be either:
- a 'Configure messages' action link that opens a dialog with the filters and clear button in there, or
- Have a filter fieldset same as the core ones (like on the User Accounts page) but make the selections cumulative, so instead of the button saying 'Filter' it would say 'Add filter', so every new selection filters down further. And a Reset button, and finally just put the clear inline with the whole lot with a confirm form.
- Or forget this till Select2 drops in core.

**Comment by jenlampton** [comment]
> Most of the time all the action you're interested in is at the top: the latest log messages.

This is exactly why we want to move the clear button out of the way: so that you can get to the thing you care about (the log messages) more easily.

> And this filter isn't same as other core filters because of the multiple

Hopefully that will be resolved by select2.

> a 'Configure messages' action link that opens a dialog with the filters and clear button

This would be dangerous. Putting a button that will delete all the things you are filtering in the same location as the filters is likely to cause people to delete the messages *by accident*. What we have now (separate fieldsets) is safer because it keeps the two different kinds of actions separated. 

> make the selections cumulative, so instead of the button saying 'Filter' it would say 'Add filter', so every new selection filters down further.

I really like this idea, but since we don't have anything like it anywhere else in core, it would be a lot of work to get us there. If we do implement it I also think we should apply this kind of filter to all the core admin views, so we still have consistent filtering throughout core.

> clear [...] with a confirm form.

I do like the idea of using a confirm form with this button. I wonder why we aren't doing that already?

**Comment by jenlampton** [comment]
I've areated another issue to add a button-danger class to that button, and filed a PR that does that, as well as moves the fieldset with the clear button above the fieldset with the filters. It sounds like we can't get consensus on what to do with the filters or the clear button yet, so I'm going to remove this milestone from *this* issue to give us some more time to figure that out :)

**Comment by quicksketch** [comment]
> I've created another issue to add a button-danger class to that button, and filed a PR that does that, as well as moves the fieldset with the clear button above the fieldset with the filters.

That was over in https://github.com/backdrop/backdrop-issues/issues/3064, which has been merged. I didn't realize there was all the contention in this issue, but I think that change no one has voiced any objection to and seemed good to me.

FWIW, I also don't much care for either of the changes suggested in this issue either.

**Comment by indigoxela** [comment]
I'm a little late here, I know. But I think, this issue deserves more attention.

Like described in a [duplicate report](https://github.com/backdrop/backdrop-issues/issues/4287), the danger of an unattended deletion of all blog entries is too high currently.

Two _very_ alike fieldsets, one does filtering, one does deletion - that's a pitfall.

Moving the button below the table (without fieldset) is a good option.
A possible alternative is a tab (like in D8). Then the action is still easy to find but the risk of an accidental deletion is significantly lower.

**Comment by indigoxela** [comment]
@opi unfortunately your PR is outdated and conflicts currently. Mind to re-roll?

@All So, let's revive this discussion. :wink: 

I created an alternate [PR](https://github.com/backdrop/backdrop/pull/3048) that puts the form in an extra tab (an extra page) - without fieldset, of course.
This moves the delete button out of the way, reducing the danger of accidental deletion.

Additionally this enables an easy way for different permissions for View and Clear (not implemented - just for possible further discussion in https://github.com/backdrop/backdrop-issues/issues/3056).

What do you think? Feedback is welcome.

**Comment by BWPanda** [comment]
Brilliant!

**Comment by laryn** [comment]
I like that solution as well. Would it make sense to reload the "View" tab after clearing the log? I know there won't be anything in there yet, but being presented with the "Clear" button (and nothing else) right after pushing it doesn't feel right.

Here are some screenshots of current PR:

![Screen Shot 2020-01-24 at 10 49 23 AM](https://user-images.githubusercontent.com/871421/73087061-7741b480-3e97-11ea-8150-092ef5c909d3.jpg)

![Screen Shot 2020-01-24 at 10 49 32 AM](https://user-images.githubusercontent.com/871421/73087065-7ad53b80-3e97-11ea-8e98-4d4b8009e04a.jpg)

**Comment by olafgrabienski** [comment]
> Would it make sense to reload the "View" tab after clearing the log? (...) being presented with the "Clear" button (and nothing else) right after pushing it doesn't feel right.

@laryn Have you seen the success message after clearing? In my opinion, with the message, it doesn't feel that wrong.

@indigoxela The success message says `Database log cleared.` I'd suggest to be slightly more verbose. Something like `The Database log has been cleared.`, or `The log messages have been cleared.`

**Comment by indigoxela** [comment]
> Would it make sense to reload the "View" tab after clearing the log?

I think, @laryn is right. Sure, the log overview is empty - but that's the desired result, so a redirect to the View page after form submit makes sense.

@olafgrabienski thanks for the message wording suggestion.

PR has been updated accordingly.

**Comment by stpaultim** [comment]
I like this!

**Comment by olafgrabienski** [comment]
Thanks @indigoxela! The updated PR works good in the sandbox site. After clearing the log messages on the *Clear* page I was redirected to the *View* page and could read the success message with updated wording.

**Comment by klonos** [comment]
I'm afraid that I see the implementation of placing the action on a different page  as a UX regression/fail, so 👎 from me. ...what we really need is #3769, but that may be a while 😅 

In the meantime, here's yet another PR, that adds a confirmation page (`confirm_form()`) to that button: https://github.com/backdrop/backdrop/pull/3055

I have moved the help text to the confirmation page, and removed the `button-danger` class from it:

![Screen Shot 2020-01-29 at 5 00 36 pm](https://user-images.githubusercontent.com/2423362/73331768-02b4ae80-42b9-11ea-95c5-ddebdd5b8387.png)

...since the "Delete" button in the confirmation is already styled as a danger button by default:

![Screen Shot 2020-01-29 at 5 00 46 pm](https://user-images.githubusercontent.com/2423362/73331765-ffb9be00-42b8-11ea-9489-b46d75bf5bef.png)

**Comment by klonos** [comment]
...I will  fix the failing tests, but waiting on some more feedback first.

For those that will be doing testing on their locals and/or code review, I could not get `'#submit' => array('dblog_clear_log_confirm')` to work as expected (it seems that the `dblog_clear_log_confirm()` was being called, but it did not return the confirmation page, and it would not trigger the db log clear). So I have used `$form['#action'] = url('admin/reports/dblog/clear')` instead. If I was doing something obviously wrong, please point it out to me, and I'll see if I can make it work with `#submit`.

**Comment by findlabnet** [comment]
After years in this discussion, maybe we can keep already workings thing as it is and focus instead it on really important tasks?

**Comment by BWPanda** [comment]
I still prefer @indigoxela's solution over @klonos'. In *my* opinion, @klonos' PR is a regression in that the clear button is now more visible than before (since the fieldset is no longer collapsible/collapsed by default).

Other pages have the delete, etc. buttons/links at the bottom, so either we should do that for consistency (but I believe there was already discussion about that option above), or move it to a separate tab as per @indigoxela's PR (IMO).

**Comment by BWPanda** [comment]
> maybe we can ... focus instead it on really important tasks?

I think if there's this much debate over an issue, it must be important to some people :wink:

**Comment by indigoxela** [comment]
@klonos I really appreciate your effort, but I think the clear button as _the very first item_ on dblog page is really a strange thing. I'm not aware of any other page/form in Backdrop that uses that design pattern.

My concern is that this might give people the impression, that clearing the logs is the main purpose of this whole page.

**Comment by klonos** [comment]
I was under the impression that the collapsible fieldset was there in order to introduce an additional click, in order to prevent accidental deletion. With a confirmation dialog in place, there'd be no point for that; but I can re-add the collaspibility, if people agree that it is a better thing UX-wise.

As for consistency this is actually what I was going for by making the fieldset not be collapsed by default. See the following UI places where a similar pattern is being used (referring to "secondary" form submit buttons):

![Screen Shot 2020-01-29 at 6 34 05 pm](https://user-images.githubusercontent.com/2423362/73336569-482ba880-42c6-11ea-8938-ecf8aa106e4c.png)

![Screen Shot 2020-01-29 at 6 28 31 pm](https://user-images.githubusercontent.com/2423362/73336572-48c43f00-42c6-11ea-86bc-5dae7baa57c8.png)

![Screen Shot 2020-01-29 at 6 27 52 pm](https://user-images.githubusercontent.com/2423362/73336574-48c43f00-42c6-11ea-9b85-4cd0ea9b958f.png)

We should aim for as much consistency in the UI as possible. So decide on whether we wrap these buttons inside fieldsets or not; decide if these fieldsets will be collapsed or not; decide whether "destructive" action buttons should be styled as "danger" buttons or not (even when there is a confirmation dialog).

At this point, I cannot think of any listing/view that has such a "delete all" button at the bottom of the page; but having said that, I am not opposed to that idea either, because I see the point of attempting to reduce prominence of that button. Would removing the fieldset entirely achieve that in a better way? ...especially since in the case of the cache clear button and this db log clear button, the fieldset labels simply repeat the label of the button:

![Screen Shot 2020-01-29 at 6 43 31 pm](https://user-images.githubusercontent.com/2423362/73336938-4dd5be00-42c7-11ea-953b-16343047c0ad.png)

![Screen Shot 2020-01-29 at 6 46 42 pm](https://user-images.githubusercontent.com/2423362/73337086-bcb31700-42c7-11ea-9c62-a92953ee2c92.png)

**Comment by klonos** [comment]
...I've updated my PR to remove the fieldset. That, combined with the removal of the `button-danger` class really tone things down.

If we have consensus re moving the button to the bottom of the page (as per the original issue report), I can then also do that. If we do that though, I really think that we should do the same with the clear cache button, for consistency.

In any way, whether we decide that the button is better at the top or the bottom of the page, I really think that it is a much better/expected UX to have a button with a confirmation page, rather than hiding it within a fieldset.

**Comment by findlabnet** [comment]
> If we have consensus 

Do not seems like consensus, IMHO.

**Comment by klonos** [comment]
> Do not seems like consensus, IMHO.

I know 😅 ...perhaps @jenlampton was right:

> Maybe we should get a ux designer to weigh in?

**Comment by BWPanda** [comment]
> I was under the impression that the collapsible fieldset was there in order to introduce an additional click, in order to prevent accidental deletion.

It may well have been, in which case the validation page serves instead. However it was also useful in minimising the scrolling needed to get to the main content of this page (what 99%* of people use this page for - viewing the log messages).

> See the following UI places where a similar pattern is being used

Cron: This makes sense as the first thing on the page, as that's what this page is all about (and it's more-or-less non-destructive)
Search indexing: As this isn't the first thing on the page, it's fine as it doesn't hinder accessing/viewing other content on the page.
Performance: I can see this one being moved further down the page, but I also don't mind it where it is as it's a more-or-less non-destructive action.

> I cannot think of any listing/view that has such a "delete all" button at the bottom of the page

No, I was thinking more about nodes, terms, etc. with Delete button at the bottom.

> Would removing the fieldset entirely achieve that in a better way?

If the button can be moved away from the top of the page (above what you're trying to access 99%* of the time), then I agree that a fieldset is not needed. But as long as the button's above the content, it should be in a collapsed fieldset.

* Made-up number

**Comment by BWPanda** [comment]
Also, I think destructive actions like irreversibly deleting all log messages should be styled appropriately (e.g. red).

**Comment by BWPanda** [comment]
All in all, my vote is still for @indigoxela's PR which:

- Reduces scrolling to content
- Avoids accidental clicking/deleting
- Removes unnecessary fieldsets

**Comment by klonos** [comment]
> * Made-up number

🤣

**Comment by findlabnet** [comment]
> trying to access 99%* of the time

Where is last one percent? :-)
Right now it is 100% of the time, for me at least.

**Comment by BWPanda** [comment]
> Where is last one percent? :-)

It's the people testing these PRs :laughing:

**Comment by olafgrabienski** [comment]
I also still like @indigoxela's PR the best. Sure, consistency is important. On the other hand, pages like *Performance* have another main purpose and are not really comparable. I'm however open for @klonos' PR if the Clear button at the top stays without fieldset and without danger class. Ironically, in that case the Clear section doesn't disturb as much as it does now.

**Comment by klonos** [comment]
FTR, my main objection here is the introduction of the UI pattern of having an action live as a secondary tab. It just does not feel right UX-wise to me.

There have been extensive discussions and user studies in Drupal-land, concluding that things that are actions (such as the "delete" operation), should be buttons/links and not tabs; tabs are being reserved for navigational operations. See:

https://www.drupal.org/project/drupal/issues/402760
https://www.drupal.org/project/drupal/issues/1834002

> Summary Summarized: this issue is just making the *location* of delete more standardized. Not a tab. Make it a action button/link in the same area as, Save, Update, Preview.
>
>(See [#7](https://www.drupal.org/project/drupal/issues/1834002#comment-6849024)) We discussed this at length in Drupal 7, and decided to go for buttons - since tabs are to be used for navigational purposes only, not actions. It is even represented in our guidelines (see the last recommendation) at Community Documentation Page: User Interface Standards: Navigation: Tabs http://drupal.org/node/1090012.

http://drupal.org/node/1090012

> Avoid using tabs for actions, there is a conceptual split in Drupal between local actions and tabs.

**Comment by docwilmot** [comment]
This conversation to me typifies my concern with our need for prioritizing in Backdrop. Two comments:

- This page works perfectly fine for its purpose. It *can* be improved, but 61 comments! And counting!
- We've started yet another new PR on an issue that had an active PR with active ongoing discussion. We could have had a conversation first, and *then* the PR if needed, no?

Excuse the rant. Carry on.

**Comment by olafgrabienski** [comment]
> This page works perfectly fine for its purpose. It can be improved, but 61 comments! And counting!

For me, it doesn't work fine at all. Every time I go to this page, it feels irritating, and I go there quite often. I see it's difficult to get consensus but that doesn't necessarily mean it's not important.

**Comment by klonos** [comment]
> We've started yet another new PR on an issue that had an active PR with active ongoing discussion.

New PRs generate new sandboxes. So people can play with both/all proposed solutions, and see which one feels better (rather than rely on mockups). I find that a good thing.

**Comment by indigoxela** [comment]
> There have been extensive discussions and user studies in Drupal-land, concluding that things that are actions (such as the "delete" operation), should be buttons/links and not tabs; tabs are being reserved for navigational operations.

@klonos ...right, and Drupal 8 uses tabs:
![screenshot-drupal8](https://user-images.githubusercontent.com/842066/73439801-7fdf3280-4350-11ea-89cd-980092d7bf8c.png)

I'm afraid, the "Curse of the Long Thread" strikes here (the longer an issue thread gets, the less likely it can get solved). Sad to see...

**Comment by BWPanda** [comment]
@klonos Regarding consistency, I see this log page as a listing of items. You can click on the items to view them (on their own page). Kind of like the content overview page (`/admin/content`), or the term listing page (`/admin/structure/taxonomy/tags`). Now imagine going to either of those pages and the first thing you see at the top is a 'Delete all [content/terms]' button... I don't think that would be very good UX, so why do that here? Let's copy D8 in this regard and have a separate tab.

**Comment by BWPanda** [comment]
> you don't go to the search page (/search) to clear the search index

Actually there's a point: the option to clear the search index in on the Search settings page. So if we're really against have a separate tab to clear the DB log, what if we move that button to the Log settings page: `/admin/config/development/logging`?

**Comment by findlabnet** [comment]
>  what if we move that button to the Log settings page

Definitely wrong place with too long path, we need clear log button on same screen where we read it.

**Comment by klonos** [comment]
> @klonos ...right, and Drupal 8 uses tabs:

I know it does (now)...

> Let's copy D8 in this regard and have a separate tab.

After doing UX reviews/studies, D8 have decided to remove all tabs and convert them to action links (again, please refer to https://www.drupal.org/project/drupal/issues/402760, where using tabs is considered UX regression, as tabs should be used for navigational purposes - not for actions). So do not rely on what is the current situation in D8, the roadmap is to change that, for the sake of better UX.

> I don't know of anywhere else in core where we make users go to the listing page in order to delete items

The way I see it, we do that everywhere actually 😅  ...and if we were to convert the list of log messages to be generated by a view (see https://www.drupal.org/project/drupal/issues/2015149 and https://www.drupal.org/node/2850115), there could be a checkbox column to allow selecting all/individual log messages, and a bulk operation to clear/delete them.

> ...we need clear log button on same screen where we read it.

**Comment by indigoxela** [comment]
Different people want different things. That happens and is usually a use-case for a contrib module.

How about that:

People who prefer to have the Clear action on a different page, install [Dblog Page Split](https://backdropcms.org/project/dblog_pagesplit)

People who prefer to have the button on the same page, proceed here with @klonos' PR.

**Comment by findlabnet** [comment]
And people who prefer to have the button as it is? 
Build another module?

**Comment by indigoxela** [comment]
> And people who prefer to have the button as it is?

I'd suggest to discuss that here. At least, some people have a (contrib) solution now. The rest is (hopefully) able to find one soon. It depends on how this discussion proceeds. :wink:

**Comment by findlabnet** [comment]
It seems like I need some module to register any core changes with the ability to roll them back.

**Comment by stpaultim** [comment]
Clearly there is substantial disagreement on this PR/issue. It's not unusual for UX related issues to get bogged down in disagreements. During last weeks meeting, we talked about this as an example of when we need some better way to make decisions. Especially, when there is no consensus. 

I am putting this on the agenda for an upcoming UX meeting.

**Comment by BWPanda** [comment]
Or should we just ask the PMC to make a final decision so we can all move on with our lives? :wink:

**Comment by findlabnet** [comment]
If you need someone who make a final decisions instead you - it is pretty simple - world of Apple/Microsoft/<another_monster> awaiting you in one step away.

**Comment by klonos** [comment]
> Or should we just ask the PMC to make a final decision so we can all move on with our lives? :wink: 

I find this issue to trivial for the PMC, but if we do take it there, then I think that perhaps I will abstain from voting, as I am directly involved (filed a PR with a solution that may come out as opinionated). ...having said that, I am currently the constituent for non-coding site builders among the PMC 😅

**Comment by BWPanda** [comment]
I don't think it's a matter of triviality, but rather do we want to spend days/weeks/months going back and forth trying to reach a consensus, or just let the PMC decide for us, commit the appropriate PR and then move on...?

**Comment by jenlampton** [comment]
@dariusgarza has agreed to be one of our our "Design Authorities" so let's bring him in here. 

The question here is: **Where should we put this "Clear Log Messages" Button?** 

There are several UX improvements we are addressing here:
* avoid accidental clicking/deleting FOR ADMINS + DEVS
* prevent confusion from too many available options above the listing FOR ADMINS + DEVS
* reduce the amount of scrolling needed to read the listing FOR ADMINS + DEVS
* make this interface more consistent with the rest of core FOR ADMINS + DEVS
* reduce #of clicks to delete messages ? FOR DEVS (only)

We will eventually want to remove the fieldset around the filters to reduce the number of clicks needed to filter the listing, too. (This will also make this interface more consistent with the rest of core). Our decision here should take that into account. 

I indicated the ROLE of the person completing each task above because in Backdrop we put the needs of ADMINS above the needs of the needs of DEVeloperS. 

Options as follows.

1) Leave it where it is, at the top of the page, buried in a fieldset:
<img width="1232" alt="Screen Shot 2020-02-20 at 10 57 54 AM" src="https://user-images.githubusercontent.com/397895/74968789-45593a80-53d0-11ea-8ea9-e9d884c7cc92.png">

2) Move it to the bottom of the page, but NOT in a fieldset: 
<img width="1606" alt="screen shot 2018-04-22 at 11 22 12 am" src="https://user-images.githubusercontent.com/6763671/39096608-56bc28d8-4620-11e8-81b2-586e47ac5e8e.png">

3) Move it to another page, accessed by a tab, like Drupal 8 did:
![screenshot-drupal8](https://user-images.githubusercontent.com/842066/73439801-7fdf3280-4350-11ea-89cd-980092d7bf8c.png)

4) Combine the two fieldsets for filtering the current list, and deleting all items to save space
(no screenshot available)

5) move that button to the Log settings page: `/admin/config/development/logging`?

### Objections to 1:

> The clear log messages is inside a fieldset. I understand that we don't want people clicking this by accident, but I'm not sure a fieldset is the way to deal with the problem

> I don't know of anywhere else in core where we make users go to the listing page in order to delete items. The delete option is always on another page/tab.

### Objections to option 2

> [DEVS will] need to scroll down this page 

### Objections to option 3

> After doing UX reviews/studies, D8 have decided to remove all tabs and convert them to action links

> introduction of [a new] UI pattern of having an action live as a primary/secondary tab. It just does not feel right UX-wise to me. 

> There have been extensive discussions and user studies in Drupal-land, concluding that things that are actions (such as the "delete" operation), should be buttons/links and not tabs; tabs are being reserved for navigational operations.

### Objections to option 4

> taking [the Clear Log Messages button] out of the fieldset and then having it anywhere near the buttons for the filters (above, below, to the right) would also be confusing [for ADMINS and DEVS]

> If the button can be moved away from the top of the page (above what you're trying to access 99%* of the time), then I agree that a fieldset is not needed. But as long as the button's above the content, it should be in a collapsed fieldset.

### Objections to option 5

>  it's convenient for DEVS to have here - right where you are anyway when you are debugging

> Definitely wrong place with too long path, we need clear log button on same screen where we read it.

I'll put this summary into the top post too...

**Comment by findlabnet** [comment]
For some reason most of programs with GUI or pseudo-GUI have user menus on top position of windows by default.

**Comment by dariusgarza** [comment]
Totally agree with @jenlampton UX overview here. Hiding the operation in a fieldset is bizarre, definitely no way should it be in a tab, also shouldn't be near the filters, and if we're worried about too much scrolling, and users accidentally nixing their data, here's a proposed solution...

![Group 2](https://user-images.githubusercontent.com/7042667/74988580-0a1c3300-53f3-11ea-850f-704ac481d7f2.png)
Put a max height on the table, or make it relative the user's browser height, and put a cap on the bottom of it in which you can place the delete button.It's also a growing heuristic to put action items like delete, submit, etc at the bottom right of a form, page, etc... Not sure if that's too far out of the design system, but may help solve some of these issues. 

![Group 3](https://user-images.githubusercontent.com/7042667/74988662-3932a480-53f3-11ea-9479-2b9db41ad2ca.png)
Present users with an "are you sure" modal to prevent any oopsies.

**Comment by laryn** [comment]
Doing it as an action button that is sticky brings to mind https://github.com/backdrop/backdrop-issues/issues/566

I like that idea.

**Comment by olafgrabienski** [comment]
I also like the idea. The `Clear Log Messages` action in the sticky element seems however very visible. Not sure if it's too eye-catching and distracting. Maybe we can mitigate that effect by styling it less prominent. Would be interesting to get an impression and be able to play around in a sandbox site.

**Comment by dariusgarza** [comment]
> Doing it as an action button that is sticky brings to mind #566
> 
> I like that idea.

That's exactly what I had in mind, great demo!

**Comment by opi** [comment]
chiming in...

> [DEVS will] need to scroll down this page

It's really NOT a valid point to me, and I'm a dev... clearing the log is not a daily task, and could be risky (aka "can't be undone"). I agree with @olafgrabienski that having a red-and-sticky button is almost a Call To Action, which I'm sure when don't want it to be.
 
The option 2 is really nice to me, without the sticky button behavior. KISS and quick win improvement.

**Comment by jenlampton** [comment]
> clearing the log is not a daily task

I agree and I'm a dev too. I also usually filter the log looking for something specific (php errors) before I clear it, and that shortens up the page a whole lot so there wouldn't be very much scrolling for me anyway.

> I agree with @olafgrabienski that having a red-and-sticky button is almost a Call To Action, which I'm sure when don't want it to be.

Hm, I hadn't thought of that You're right that it might seem like a "save" button or something people are supposed to do here, and might increase the probability of people doing it by accident. Would it have the same feel when it's just placed at the bottom of the page, but not in a sticky area?

I'm not sure getting the sticky buttons in is a realistic end-goal for this particular issue / PR anyway since we should really put in something reusable for #566 if/when we do that. 

We have also talked about modal confirm forms in https://github.com/backdrop/backdrop-issues/issues/3769 so it sounds to me like this is 3 separate issues:

* https://github.com/backdrop/backdrop-issues/issues/3769
* #566 w a reusable pattern we can apply here
* Move the button to the bottom of this page (maybe not in a fieldset?)

@dariusgarza do you have any recommendations for this intermediate step when the button is at the bottom of the page but not yet in a sticky box...  should the fieldset be collapsed to hide the button, or not?

**Comment by docwilmot** [comment]
My brief input: I feel these two are not much different:
Content page:
![content](https://user-images.githubusercontent.com/1042283/75077780-b7657880-54d1-11ea-9a87-1905dcae3652.png)

The log page:
![Recent log messages   backdrop localhost](https://user-images.githubusercontent.com/1042283/75077788-c0eee080-54d1-11ea-8a87-3b515a1707d2.png)

The button is **red** and there is a confirm form; if a user accidentally clears the form... well there is only so much hand-holding we should be doing.

This is not a form with sequential inputs ending in an action like Save or Configure, so putting the button at the end is not appropriate. 

The recommendation by @dariusgarza looks good but we dont do this anywhere else.

**Comment by jenlampton** [comment]
This page doesn't have bulk operations like the content, user, or files
listing. The "delete all" button is not a parallel to what you see in that
screenshot. In order to delete all items you'd need to intentionally do two
other things first: 1 select all items, and 2 select the delete action.
This form has neither of those so shouldn't be compared from a user
experience standpoint. Yes, there's a point where they look the same, but
the use of them isn't the same at all.

**Comment by olafgrabienski** [comment]
> The button is red and there is a confirm form

@docwilmot Are you referring to one of the PRs? In current Backdrop, there is no confirmation step. If you push the red button, the log is cleared right away.

**Comment by olafgrabienski** [comment]
> having a red-and-sticky button is almost a Call To Action

And IIRC, Backdrop makes only use of buttons with red outlines (hover: full red) just in the moment before actions can't be undone, e.g. on the confirmation page before deleting a node. In contrast, the `Delete` button on the content edit page is styled without danger class, just with a gray outline.

I guess, historically the button on the DBlog page is rightly red, because there is no confirmation step. Now, when we introduce the confirmation step (which we should), we don't need the red color anymore.

**Comment by klonos** [comment]
@olafgrabienski's last comment sums up my opinion re solving the "accidental deletion" possibility, that others have brought up: we should add a confirmation page, and move the button outside the fieldset. My PR in https://github.com/backdrop/backdrop/pull/3055 does that, so I would love some feedback on whether the workflow makes sense.

I realize that we still have objections re the position of the button, but I am open to suggestions/change re that. My point is: let's get the button out of the fieldset (fixes UI weirdness) and add a confirmation (fixes data loss hazard); let's just decide on the button position and get this done 🙂 

... @jenlampton's mentions of #3769 and #566 sum up my opinion re next steps. These will not happen as part of this issue here, but once these 2 other features have been implemented, we can apply them here.

Thank you @dariusgarza for taking the time to look into this and chime in; your professional opinion in this is greatly appreciated 👍

**Comment by jenlampton** [comment]
@klonos I tested this PR

<img width="1245" alt="Screen Shot 2020-02-27 at 11 58 23 AM" src="https://user-images.githubusercontent.com/397895/75481851-8c0edd80-5958-11ea-9f99-88b4d1bc3f60.png">

I am worried that particular configuration might make it more likely for people to delete log messages, even with the confirm form. 
1) we removed the button-danger class
2) we also removed the text from the main page telling people this was dangerous
3) it's now the first thing at the top on the page, and looks friendly

I wonder if it might better to leave the fieldset (but expanded, to avoid the extra click) with the warning text as it was. I'd like to see a version of this that looks more like #2 in the examples above, but we can remove the button-danger class, and leave the confirm form.

**Comment by olafgrabienski** [comment]
After some time, I've had a fresh look at the PR of @klonos, and the UI looks really good to me. The way I see it, the PR won't make it more likely to delete log messages unintentionally.

1. Removed button-danger class makes it *less* likely to click at the button in this case.
2. Removed text (see next line) may make a difference. I'd be okay to add it again.
  `This action will permanently delete the log messages from the database.`

3. Same as for 1, in my opinion.

**Comment by klonos** [comment]
This was discussed during the weekly UX/design meeting. How does everyone feel about the following options:

- Escalate this to the Project Management Committee, so that a decision is made.
- Park this, until we have #566 and #3769 in core, at which point we could implement the designs proposed by @dariusgarza in https://github.com/backdrop/backdrop-issues/issues/2353#issuecomment-589401931

**Comment by klonos** [comment]
FTR, @jenlampton I disagree with your last review. Take the "Delete" button in the node edit form for example:

![image](https://user-images.githubusercontent.com/2423362/91499490-2cbf6500-e8ca-11ea-8b50-6f82c3a35418.png)

...that also looks "friendly", and we don't have any specific "danger" indication or text accompanying it. There's a dialog with a danger button and text that explains things:

![image](https://user-images.githubusercontent.com/2423362/91500633-89bc1a80-e8cc-11ea-8ef3-807ca45ddedf.png)

And this pattern is basically the same with canceling user accounts, deleting content types, layouts, menus, views etc. I am not doing anything differently in the current PR 🤷 

One thing that we could do in order to avoid accidental clicks, is to float the button to the right (left on RTL languages):

![image](https://user-images.githubusercontent.com/2423362/91501154-b45aa300-e8cd-11ea-933e-d4f003675f1c.png)

That way, it's not "visually grouped" with the main-focus elements on the page (breadcrumb, page title, filter fieldset label etc.), and by it having no "danger" style, it is also not "screaming" for attention.

**Comment by jenlampton** [comment]
But there's no confirm form on this page, and there are on those pages.
Those pages need one because they delete important things like content,
this page doesn't need one, so shouldn't we use a more dangerous button
color?

**Comment by BWPanda** [comment]
I vote for escalating this to the PMC, then they can decide what direction to take (parking this, top or bottom of the screen, danger class or not, etc.).

**Comment by olafgrabienski** [comment]
> But there's no confirm form on this page, and there are on those pages.

Clarifying before we continue in the PMC: I've just had another look at the [PR sandbox site](http://3055.backdrop.backdrop.qa.backdropcms.org). It (rightly) *does* provide a confirm form.

**Comment by BWPanda** [comment]
I updated a comment above that I made earlier this year, but I'll add it here too, since there seems to be some confusion about when to add the 'danger' class (i.e. red button) and when to have a confirmation form.

As far as I can tell, the pattern Backdrop uses is like this:

- If clicking a button will perform a destructive action straight away, then that button is given the 'danger' class (i.e. styled red). For example, confirmation forms when deleting a node, etc.
- If clicking a button displays a confirmation form, then technically *that* button is not destructive and is therefore styled normally. For example, the Delete button at the bottom of node edit forms.

The current situation for the 'Clear all log messages' button is that there is no confirmation form and therefore the button is rightly given the 'danger' class (styled red).

Some people are proposing that a confirmation form be added, in which case this 'Clear all log messages' button will be styled normally (i.e. grey) and the new button on the new confirmation form will be styled red.

The confusion seems to be stemming from screenshots of the new, grey button. Since screenshots don't show the addition of a confirmation form, people are thinking we're removing the 'danger' class without knowing that a confirmation form (with a red button) is *also* being added/proposed.

I hope that clears that up.

**Comment by BWPanda** [comment]
In today's [Design/UX meeting](https://www.youtube.com/watch?v=XvC9ZUlVR-M) this issue was discussed and PMC members in the meeting seemed to suggest that they would need a summary of this issue before they discuss/vote on it. I've therefore gone ahead and made the following summary after re-reading through this issue:

---

## Problems

1. Destructive action button too prominent at top of page
2. Need to scroll past button to get to log listing
3. Additional click(s) needed to clear log messages

## Solutions

1. **Leave it as-is**
   Problems fixed: None
   Thoughts:
     - Fieldset not the best solution for avoiding accidental button clicks
     - Elsewhere in core, the delete button for a listing is on another page/tab
     
2. **Move button to bottom of page, no fieldset**
   Problems fixed: 1, 2, 3
   Thoughts:
     - Need to scroll down the page to clear messages
     - Might be seen as an action button (i.e. made more prominent)

3. **Move button to a new tab**
   Problems fixed: 1, 2
   Thoughts:
     - This is how Drupal 8 does it, but they're in the process of changing it (based on research/studies)
     - New UI pattern (we don't do this elsewhere in core)

4. **Move button into Filter fieldset (and have it expanded)**
   Problems fixed: 2, 3
   Thoughts:
     - Might be more confusing having it near filter button(s)
     - Button should not be immediately visible above listing (i.e. fieldset should be collapsed)

5. **Move button to Logging and Errors page**
   Problems fixed: 1, 2
   Thoughts:
     - Not as convenient, moving button to separate section of the site

6. **Move button to floating section at bottom of screen, display confirmation in modal**
   Problems fixed: 1, 2, 3
   Thoughts:
     - 'Sticky' button might be too visible/prominent
     - Not consistent with rest of UI (unless https://github.com/backdrop/backdrop-issues/issues/566 is implemented)

---

I hope this is sufficient for the PMC to make a decision on this.

**Comment by findlabnet** [comment]
IMHO, to make the system more attractive to the end user, we don't need endless movement of buttons or polish of the tongue, but instead we need something really valuable - for example, **correctly working** paragraphs (available right out of the box.)

**Comment by indigoxela** [comment]
> to make the system more attractive to the end user, we don't need endless movement of buttons or polish of the tongue, but instead we need something really valuable - for example, correctly working paragraphs

@findlabnet although this endless discussion is hard to follow, this issue deals with a _real_ problem: two very alike buttons (fieldsets actually), next to each other. In *switched* order compared to Drupal, one of them with a destructive action. That's an unnecessary pitfall.

It's simply not helpful, if you try to divert us with something unrelated, for instance paragraphs. That whataboutism is frustrating for people, who try to contribute something useful to get some progress here. If you don't have the problem, this issue tries to solve, fine. But please accept that other people *do* have a problem with it.

**Comment by findlabnet** [comment]
Sorry for my so wrong opinion. 
Will not participate in this thread anymore.

**Comment by olafgrabienski** [comment]
> summary after re-reading through this issue / https://github.com/backdrop/backdrop-issues/issues/2353#issuecomment-738498521

@BWPanda Thanks for the concise summary, it'll help a lot. There's one topic missing, in my opinion: the discussion about adding a confirmation form (and changing the button color). You have however already summarized that topic nicely, just [a few comments above](https://github.com/backdrop/backdrop-issues/issues/2353#issuecomment-738496133) in the thread. I guess, linking to that comment might be enough to give an idea.

**Comment by BWPanda** [comment]
@olafgrabienski I think the main decision is *where* the button should be displayed. I think a separate decision is if it should be converted to a regular button that has a confirmation form. But I'll let the PMC decide on all that.

In other news, I just realised that we have a similar situation with the Simpletest UI: `admin/config/development/testing` - on that page, the list of tests comes first, then there's a 'clean environment' button at the bottom... In case that helps make a decision about keeping consistency in core's UI.

**Comment by klonos** [comment]
> ...we have a similar situation with the Simpletest UI

Good catch @BWPanda 👍 ...and both these pages are meant for the same audience (site admins/developers).

**Comment by stpaultim** [comment]
I have officially opened a ticket in the PMC issue queue for this issue. I expect it will take at least several weeks for any resolution. If anyone has ideas on how to resolve this without the PMC, I'm sure that the PMC would be fine with that. :-)

**Comment by BWPanda** [comment]
Status report...?

**Comment by BWPanda** [comment]
Seriously, what's the status of this issue?

**Comment by klonos** [comment]
@BWPanda last I recall is that there's various options/opinions here, but we need a UX/design opinion from an expert (hence the `design` label). Basically we need to make a decision.

**Comment by BWPanda** [comment]
> we need a UX/design opinion from an expert

@klonos I thought that's what this was: https://github.com/backdrop/backdrop-issues/issues/2353#issuecomment-589401931

**Comment by olafgrabienski** [comment]
As requested, the PMC talked about it already last year, but we didn't manage to communicate the result for various reasons. (Sorry for that!) One of the reasons: the PMC discussion result isn't absolutely clear. There was a preference for two of the approaches mentioned in your summary https://github.com/backdrop/backdrop-issues/issues/2353#issuecomment-738498521.

As a preparation, I list the different approaches from the [summary](https://github.com/backdrop/backdrop-issues/issues/2353#issuecomment-738498521) again:

1. Leave it as-is
2. Move button to bottom of page, no fieldset
3. Move button to a new tab
4. Move button into Filter fieldset (and have it expanded)
5. Move button to Logging and Errors page
6. Move button to floating section at bottom of screen, display confirmation in modal

Our impression of the discussion in this very issue thread was: At a certain point, most support was expressed for solution 3, moving the button to a new tab. But after solution 3 became available as a contrib module, the discussion became more open toward solution 2 again. At this point, the discussion stopped.

The PMC itself remained undecided between approach 2 and 3, but all members like one both. At the end, we didn't want to decide but to give the question back to the community, pointing to the two mentioned approaches.

**Comment by jenlampton** [comment]
Thanks @olafgrabienski !  The last issue in the PMC queue recommends that we promote option #2.

**Comment by BWPanda** [comment]
Thanks @olafgrabienski and @jenlampton. Nice to have an update on this issue and a way to move forward.

I'd like to recommend now that we close this (very long) issue and open a new one where we can make a PR to test and develop this change.

**Comment by jenlampton** [comment]
I think that's a fantastic idea @BWPanda :) thank you!  Would you do the honor of creating the next issue?

**Comment by BWPanda** [comment]
Here's the follow-up issue for anyone interested: https://github.com/backdrop/backdrop-issues/issues/5484
